include/pybind11/cast.h

```diff
@@ -2204,6 +2204,18 @@ object object_api<Derived>::call(Args &&...args) const {
 
 PYBIND11_NAMESPACE_END(detail)
 
+
+template<typename T>
+type type::of() {
+   static_assert(
+      std::is_base_of<detail::type_caster_generic, detail::make_caster<T>>::value,
+      "py::type::of<T> only supports the case where T is a registered C++ types."
+    );
+
+    return type((PyObject*) detail::get_type_handle(typeid(T), true).ptr(), borrowed_t());
```

Comment:
Ahhh, nuts!
I forgot to ask: Would you be OK with exposing a "type erased" version of this?
e.g. `py::type::of(const std::type_info&)`?

It would be useful for us in code that mentioned before, e.g. here:
https://github.com/RobotLocomotion/drake/blob/a7e2a49cb673e73c40a327478850f56a3e5fde9d/bindings/pydrake/common/cpp_param_pybind.cc#L102-L111

I can do it as a follow-up if you'd like to get this merged in!

Reply:
This would add a new stdlib import?

Reply:
Are we sure about releasing these implementation details?
Is there a reason you can't make this a template function in your client code, @EricCousineau-TRI?

Then again, I guess these implementation details are not bound to change too soon.

Reply:
This also might add complications if you expect `typeid(int)` or so to work, though?

Banning this "power-use" to the `detail` namespace, and not publicly exposing it (and potentially getting bugs, etc) - i.e., what we have now - could just work as well. This `detail` might not soon change, but I'm wondering if it's a good idea to add to the public API :-/

Reply:
I think he's just asking for a runtime version of `of<T>` using `std::type_info`? I don't think that's exposing an implementation detail?

Reply:
Given that's what's returned by `typeid(T)` anyway, it should not be an extra dependency, I can add that, I think. And one can be based on the other with no duplication.

Reply:
Okay, maybe eventually then. I think this is cleared to merge, then?

Reply:
OK, we can see. Just wanted to express a few doubts; not veto'ing anything. But let's get that in another PR, then, afaic :-)